src/gwt/src/org/rstudio/studio/client/workbench/views/source/Source.java

```diff
@@ -538,6 +544,88 @@ public void onExecuteUserCommand(ExecuteUserCommandEvent event)
       events.addHandler(DocWindowChangedEvent.TYPE, this);
       events.addHandler(DocTabDragInitiatedEvent.TYPE, this);
       events.addHandler(PopoutDocInitiatedEvent.TYPE, this);
+      
+      events.addHandler(
+            ReplaceSelectionEvent.TYPE,
+            new ReplaceSelectionEvent.Handler()
+            {
+               @Override
+               public void onReplaceSelection(ReplaceSelectionEvent event)
+               {
+                  if (activeEditor_ == null)
+                     return;
+                  
+                  if (!(activeEditor_ instanceof TextEditingTarget))
+                     return;
+                  
+                  TextEditingTarget target = (TextEditingTarget) activeEditor_;
+                  target.insertCode(event.getData(), false);
+               }
+            });
+      
+      events.addHandler(
+            ReplaceRangesEvent.TYPE,
+            new ReplaceRangesEvent.Handler()
+            {
+               @Override
+               public void onReplaceRanges(ReplaceRangesEvent event)
+               {
+                  if (activeEditor_ == null)
+                     return;
+                  
+                  if (!(activeEditor_ instanceof TextEditingTarget))
+                     return;
+                  
+                  TextEditingTarget target = (TextEditingTarget) activeEditor_;
+                  
+                  JsArray<ReplacementData> data = event.getData().getReplacementData();
+                  for (int i = 0; i < data.length(); i++)
+                  {
+                     ReplacementData el = data.get(i);
+                     Range range = el.getRange();
+                     String text = el.getText();
+                     
+                     target.getDocDisplay().replaceRange(range, text);
+                  }
+               }
+            });
+      
+      events.addHandler(
+            GetActiveDocumentContextEvent.TYPE,
```

Comment:
This was a bit trickier to handle than expected, as the default mode is for the server to fire events to the main window + all satellites. In this case, I needed to accept the event only on the main window, and then dispatch a separate event to the last focused window. This is currently accomplished by creating separate 'dispatch' event which do not forward back to the main window, and then firing a separate event to the satellite (or main window) which should actually receive the event.

Is it possible for us to control whether a client event emitted from the server is targeted only to the main window?